File: solutions/ocaml/123-sliding-window-maximum.ml
(*
  Problem 123: Sliding Window Maximum (LeetCode 239)
  Difficulty: Hard
  Language: OCaml
*)
let max_sliding_window nums k =
  let n = Array.length nums in
  let rec go i dq res =
    if i >= n then List.rev res
    else
      let dq = match dq with j :: _ when j <= i - k -> List.tl dq | _ -> dq in
      let rec pop_back = function
        | [] -> []
        | [j] when nums.(j) <= nums.(i) -> []
        | l -> let rev = List.rev l in
          if nums.(List.hd rev) <= nums.(i) then List.rev (List.tl rev) |> pop_back
          else l in
      let dq = pop_back dq @ [i] in
      let res = if i >= k - 1 then nums.(List.hd dq) :: res else res in
      go (i + 1) dq res
  in go 0 [] []

File: solutions/rust/123-sliding-window-maximum.rs
/**
 * Problem 123: Sliding Window Maximum (LeetCode 239)
 * Difficulty: Hard
 * Language: Rust
 */
use std::collections::VecDeque;
fn max_sliding_window(nums: &[i32], k: usize) -> Vec<i32> {
    let (_, res) = nums.iter().enumerate().fold(
        (VecDeque::new(), vec![]),
        |(mut dq, mut res), (i, &v)| {
            while dq.front().map_or(false, |&j| j + k <= i) { dq.pop_front(); }
            while dq.back().map_or(false, |&j| nums[j] <= v) { dq.pop_back(); }
            dq.push_back(i);
            if i >= k - 1 { res.push(nums[*dq.front().unwrap()]); }
            (dq, res)
        });
    res
}
